[Merged by Bors] - feat(algebra/lie/weights): define product of root vectors and weight vectors

ocfnash opened the pull request:
Also some related results, most notably that the zero root space is a subalgebra.

---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/lie/weights): define product of root vectors and weight vectors

Also some related results, most notably that the zero root space is
a subalgebra.

ocfnash commented on src/algebra/lie/weights.lean:
```diff
@@ -231,37 +231,142 @@ end lie_module
+    subtype.coe_mk, subtype.coe_mk],
+  map_smul' := λ t x, by ext m; rw [linear_map.smul_apply, linear_map.coe_mk, linear_map.coe_mk,
+    subtype.coe_mk, lie_submodule.coe_smul, smul_lie, lie_submodule.coe_smul, subtype.coe_mk],
+  map_lie'  := λ x y, by ext m; rw [lie_hom.lie_apply, lie_submodule.coe_sub, linear_map.coe_mk,
```
For reasons I don't understand I must use `rw` in most of the proof obligations of this definition.

For example if I replace the `rw` here with `simp only` then I get a deterministic timeout (even with a very long time limit). In fact even taking just the first two terms of this `rw`, namely `simp only [lie_hom.lie_apply, lie_submodule.coe_sub]` times out (after first consuming more and more memory).

Is there a simp loop? I don't really know how to debug this.

ocfnash replied:
@kbuzzard  just pointed out to me the problems with `simp` are now gone on current master. Indeed the following works:
```lean
def root_space_weight_space_product (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃) :
  (root_space H χ₁) ⊗[R] (weight_space M χ₂) →ₗ⁅R,H⁆ weight_space M χ₃ :=
lift_lie R H (root_space H χ₁) (weight_space M χ₂) (weight_space M χ₃)
{ to_fun    := λ x,
  { to_fun    :=
      λ m, ⟨⁅(x : L), (m : M)⁆,
            hχ ▸ (lie_mem_weight_space_of_mem_weight_space x.property m.property) ⟩,
    map_add'  := λ m n, by simpa,
    map_smul' := λ t m, by simpa, },
  map_add'  := λ x y, by ext m; simp,
  map_smul' := λ t x, by ext m; simp,
  map_lie'  := λ x y, by ext m; simp, }
```
However I will leave the implementation as it stands because elaboration time on my machine increases from about 3.4s to 12.8s if I switch to the above.

It is a mystery as to what the problem was but I can reproduce it by navigating to the commit edf56dfce079f0732084fda74073ecd6c19229d2 on this (temporarily restored) branch.

Using `set_option trace.simplify.rewrite true` I see no loop and using the more verbose `set_option trace.simplify true`, I see nothing obvious, but I do not understand its output well.

As regards why this works on master but not on the old commit, I tested and it appears to be the result of the changed definition from:
```lean
def root_space_weight_space_product (χ₁ χ₂ : H → R) :
```
to:
```lean
def root_space_weight_space_product (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃) :
```
that was added (for other reasons) in the final commit of this PR.

ocfnash committed:
Tweak variable naming

eric-wieser commented on src/algebra/lie/subalgebra.lean:
```diff
@@ -137,12 +138,23 @@ instance : lie_ring_module L' M :=
+`L' ⊆ L`. -/
+def restrict_lie_module_hom (f : M →ₗ⁅R,L⁆ N) : M →ₗ⁅R,L'⁆ N :=
+{ map_lie' := λ x m, f.map_lie ↑x m,
+  .. (f : M →ₗ[R] N)}
```
We already have `linear_map.restrict_scalars` which means something similar - so maybe this should be called `lie_module_hom.restrict_lie` or `lie_module_hom.restrict_lie_ring` both to enable dot notation and for consistency

ocfnash replied:
Agreed. I've just added https://github.com/leanprover-community/mathlib/pull/7591/commits/b132960c36f4e2e016d92a3011423a885b56b920 to cater for this.

ocfnash committed:
Rename `lie_subalgebra.restrict_lie_module_hom` → `lie_module_hom.restrict_lie`

As pointed out by Eric, this is more consistent with `linear_map.restrict_scalars`

ocfnash committed:
Minor comment tweaks

jcommelin commented on src/algebra/lie/weights.lean:
```diff
@@ -231,37 +233,142 @@ end lie_module
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L` together with `χ₁ χ₂ : H → R`, there is a natural
+`R`-linear product of root vectors, compatible with the actions of `H`. -/
+def root_space_product (χ₁ χ₂ : H → R) :
```
One thing that might turn out to be really annoying is that you know have `χ₁ + χ₂` in the target, which is not defeq to `χ₂ + χ₁`. Similarly if you chain these products together, you will get associativity issues, etc. I guess this already happens with `root_space_weight_space_product` as well.

jcommelin replied:
I guess there are at least two approaches around this:

1. Take a third argument `χ3` and a proof `h : χ₁ + χ₂ = χ3`
2. Only take the `χ3` and just return `0` if `χ₁ + χ₂ = χ3` is false. (This can be justified by saying that `root_space_product χ₁ χ₂ χ3` is the composition of "inject into the graded algebra of root spaces, take the product, and project down to root space `χ3`.)

ocfnash replied:
Hmm, very good point. Option 2 is in particular is tempting but at the very least I think I should look more closely at the approach to gradings that @eric-wieser and @kbuzzard have developed as part of LTE.

I've committed (to myself) to spend today refereeing an article but I'll try to give this some thought in parallel.

ocfnash replied:
I have now adjusted this (choosing option 1 in the end).

ocfnash committed:
Change `root_space_weight_space_product` to allow explicit control of target type

jcommelin reviewed (approved):
Thanks :tada:

bors merge
